// react-sketchapp/lib/index.js
"use strict";

var _render = require("./render");

var _Platform = require("./Platform");

var _stylesheet = require("./stylesheet");

var _Document = require("./components/Document");

var _Page = require("./components/Page");

var _Artboard = require("./components/Artboard");

var _Image = require("./components/Image");

var _RedBox = require("./components/RedBox");

var _Svg = require("./components/Svg");

var _View = require("./components/View");

var _Text = require("./components/Text");

var _TextStyles = require("./sharedStyles/TextStyles");

var _symbol = require("./symbol");

module.exports = {
  render: _render.render,
  renderToJSON: _render.renderToJSON,
  StyleSheet: _stylesheet.default,
  Document: _Document.default,
  Page: _Page.default,
  Artboard: _Artboard.default,
  Image: _Image.default,
  RedBox: _RedBox.default,
  Svg: _Svg.default,
  Text: _Text.default,
  TextStyles: _TextStyles.default,
  View: _View.default,
  Platform: _Platform.default,
  makeSymbol: _symbol.makeSymbol,
  getSymbolComponentByName: _symbol.getSymbolComponentByName,
  injectSymbols: _symbol.injectSymbols
};
