// react-sketchapp/lib/sharedStyles/TextStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var invariant = require("invariant");

var _sketchappJsonPlugin = require("sketchapp-json-plugin");

var _hashStyle = require("../utils/hashStyle");

var _sharedTextStyles = require("../wrappers/sharedTextStyles");

var _hacksForJSONImpl = require("../jsonUtils/hacksForJSONImpl");

var _pick = require("../utils/pick");

var _constants = require("../utils/constants");

var _styles = {};
var _byName = {};

var registerStyle = function registerStyle(name, style) {
  var safeStyle = (0, _pick.default)(style, _constants.INHERITABLE_FONT_STYLES);
  var hash = (0, _hashStyle.default)(safeStyle);
  var sketchStyle = (0, _hacksForJSONImpl.makeTextStyle)(safeStyle);

  var sharedObjectID = _sharedTextStyles.default.addStyle(name, sketchStyle); // FIXME(gold): side effect :'(


  _byName[name] = hash;
  _styles[hash] = {
    cssStyle: safeStyle,
    name: name,
    sketchStyle: sketchStyle,
    sharedObjectID: sharedObjectID
  };
};

var create = function create(options, styles) {
  var clearExistingStyles = options.clearExistingStyles,
      context = options.context;

  if (!(0, _sketchappJsonPlugin.appVersionSupported)()) {
    context.document.showMessage('💎 Requires Sketch 43+ 💎');
    return {};
  }

  invariant(options && options.context, 'Please provide a context');

  _sharedTextStyles.default.setContext(context);

  if (clearExistingStyles) {
    _styles = {};

    _sharedTextStyles.default.setStyles([]);
  }

  Object.keys(styles).forEach(function (name) {
    return registerStyle(name, styles[name]);
  });
  return _styles;
};

var resolve = function resolve(style) {
  var hash = (0, _hashStyle.default)(style);
  return _styles[hash];
};

var get = function get(name) {
  var hash = _byName[name];
  var style = _styles[hash];
  return style ? style.cssStyle : {};
};

var clear = function clear() {
  _styles = {};

  _sharedTextStyles.default.setStyles([]);
};

var styles = function styles() {
  return _styles;
};

var TextStyles = {
  create: create,
  resolve: resolve,
  get: get,
  styles: styles,
  clear: clear
};
var _default = TextStyles;
exports.default = _default;
