// react-sketchapp/lib/jsonUtils/style.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeHorizontalBorder = exports.makeVerticalBorder = exports.makeShadow = exports.makeBorderOptions = void 0;

var _sketchConstants = require("sketch-constants");

var _models = require("./models");

var _shapeLayers = require("./shapeLayers");

var DEFAULT_SHADOW_COLOR = '#000';

var makeDashPattern = function makeDashPattern(style, width) {
  switch (style) {
    case 'dashed':
      return [width * 3, width * 3];

    case 'dotted':
      return [width, width];

    case 'solid':
      return [];

    default:
      return [];
  }
};

var makeBorderOptions = function makeBorderOptions(style, width) {
  return {
    _class: 'borderOptions',
    isEnabled: false,
    dashPattern: makeDashPattern(style, width),
    lineCapStyle: 0,
    lineJoinStyle: 0
  };
};

exports.makeBorderOptions = makeBorderOptions;

var makeShadow = function makeShadow(style) {
  var opacity = style.shadowOpacity !== undefined ? style.shadowOpacity : 1;
  var color = style.shadowColor || style.color || DEFAULT_SHADOW_COLOR;
  var radius = style.shadowRadius !== undefined ? style.shadowRadius : 1;

  var _class = style.shadowInner !== undefined ? 'innerShadow' : 'shadow';

  var spread = style.shadowSpread !== undefined ? style.shadowSpread : 0;

  var _ref = style.shadowOffset || {},
      _ref$width = _ref.width,
      offsetX = _ref$width === void 0 ? 0 : _ref$width,
      _ref$height = _ref.height,
      offsetY = _ref$height === void 0 ? 0 : _ref$height;

  return {
    _class: _class,
    isEnabled: true,
    blurRadius: radius,
    color: (0, _models.makeColorFromCSS)(color, opacity),
    contextSettings: {
      _class: 'graphicsContextSettings',
      blendMode: 0,
      opacity: 1
    },
    offsetX: offsetX,
    offsetY: offsetY,
    spread: spread
  };
};

exports.makeShadow = makeShadow;

var makeVerticalBorder = function makeVerticalBorder(x, y, length, thickness, color) {
  var frame = (0, _models.makeRect)(x, y, thickness, length);
  var shapeFrame = (0, _models.makeRect)(0, 0, thickness, length);
  var shapePath = (0, _shapeLayers.makeShapePath)(shapeFrame, (0, _shapeLayers.makeVerticalPath)());
  var content = (0, _shapeLayers.makeShapeGroup)(frame, [shapePath]);
  content.style.borders = [{
    _class: 'border',
    isEnabled: true,
    color: (0, _models.makeColorFromCSS)(color),
    fillType: 0,
    position: _sketchConstants.BorderPosition.Center,
    thickness: thickness
  }];
  return content;
};

exports.makeVerticalBorder = makeVerticalBorder;

var makeHorizontalBorder = function makeHorizontalBorder(x, y, length, thickness, color) {
  var frame = (0, _models.makeRect)(x, y, length, thickness);
  var shapeFrame = (0, _models.makeRect)(0, 0, length, thickness);
  var shapePath = (0, _shapeLayers.makeShapePath)(shapeFrame, (0, _shapeLayers.makeHorizontalPath)());
  var content = (0, _shapeLayers.makeShapeGroup)(frame, [shapePath]);
  content.style.borders = [{
    _class: 'border',
    isEnabled: true,
    color: (0, _models.makeColorFromCSS)(color),
    fillType: 0,
    position: _sketchConstants.BorderPosition.Center,
    thickness: thickness
  }];
  return content;
};

exports.makeHorizontalBorder = makeHorizontalBorder;
