// react-sketchapp/lib/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.APPLE_BROKEN_SYSTEM_FONT = exports.VALID_TEXT_CHILDREN_TYPES = exports.INHERITABLE_FONT_STYLES = exports.PatternFillType = void 0;

/* eslint-disable import/prefer-default-export */
// out of date in sketch-constants
// https://github.com/turbobabr/sketch-constants/pull/1
var PatternFillType = {
  Tile: 0,
  Fill: 1,
  Stretch: 2,
  Fit: 3
};
exports.PatternFillType = PatternFillType;
var INHERITABLE_FONT_STYLES = ['color', 'fontFamily', 'fontSize', 'fontStyle', 'fontWeight', 'textAlign', 'textDecoration', 'textShadowOffset', 'textShadowRadius', 'textShadowColor', 'textTransform', 'letterSpacing', 'lineHeight', 'writingDirection']; // Only components that are allowed as children of <Text> components

exports.INHERITABLE_FONT_STYLES = INHERITABLE_FONT_STYLES;
var VALID_TEXT_CHILDREN_TYPES = ['text']; // Font displayed if San Francisco fonts are not found

exports.VALID_TEXT_CHILDREN_TYPES = VALID_TEXT_CHILDREN_TYPES;
var APPLE_BROKEN_SYSTEM_FONT = '.AppleSystemUIFont';
exports.APPLE_BROKEN_SYSTEM_FONT = APPLE_BROKEN_SYSTEM_FONT;
